tests/test_qp.py

```diff
@@ -22,317 +24,245 @@ class QPTest(PyverbsAPITestCase):
-        Checked QP types are Raw Packet and UD. Raw Packet is skipped for
-        non-root users / Infiniband link layer.
-        """
+    def create_qp(self, creator, qp_init_attr, is_ex, with_attr):
+        try:
+            qp_attr = (None, QPAttr())[with_attr]
+            return QP(creator, qp_init_attr, qp_attr)
+        except PyverbsRDMAError as ex:
+            if ex.error_code == errno.EOPNOTSUPP:
+                with_str = ('with', 'without')[with_attr] + ('', ' extended')[is_ex]
+                raise unittest.SkipTest(f'Create {get_qp_type_str(qp_init_attr.qp_type)} QP {with_str} attrs is not supported')
+            raise ex
+
+    def create_qp_common_test(self, qp_type, qp_state, is_ex, with_attr):
         for ctx, attr, attr_ex in self.devices:
             with PD(ctx) as pd:
                 with CQ(ctx, 100, None, None, 0) as cq:
-                    for i in range(1, attr.phys_port_cnt + 1):
+                    if qp_type == e.IBV_QPT_RAW_PACKET:
+                        eth_port = 0
+                        for i in range(1, attr.phys_port_cnt + 1):
+                            if u.is_eth(ctx, i) and u.is_root():
+                                eth_port = i
+                                break
+                        if eth_port == 0:
+                            raise unittest.SkipTest('To Create RAW QP must done by root on Ethernet link layer')
+
+                    if is_ex:
+                        qia = get_qp_init_attr_ex(cq, pd, attr, attr_ex, qp_type)
+                        creator = ctx
+                    else:
                         qia = get_qp_init_attr(cq, attr)
-                        qia.qp_type = e.IBV_QPT_UD
-                        with QP(pd, qia) as qp:
-                            assert qp.qp_state == e.IBV_QPS_RESET, 'UD QP should have been in RESET'
-                        if u.is_eth(ctx, i) and u.is_root():
-                            qia.qp_type = e.IBV_QPT_RAW_PACKET
-                            try:
-                                with QP(pd, qia) as qp:
-                                    assert qp.qp_state == e.IBV_QPS_RESET, 'Raw Packet QP should have been in RESET'
-                            except PyverbsRDMAError as ex:
-                                if ex.error_code == errno.EOPNOTSUPP:
-                                    raise unittest.SkipTest("Create Raw Packet QP is not supported")
-                                raise ex
-
-    def test_create_qp_with_attr_connected(self):
-        """
-        Test QP creation via ibv_create_qp without a QPAttr object proivded.
-        Checked QP types are RC and UC.
+                        qia.qp_type = qp_type
+                        creator = pd
+
+                    qp = self.create_qp(creator, qia, is_ex, with_attr)
+                    assert qp.qp_state == qp_state , f'{get_qp_type_str(qp_type)} QP should have been in {get_qp_state_str(qp_state)}'
```

Comment:
please assign "get_qp_type_str(qp_init_attr.qp_type)" to a temp var. (128c length is just too much).
remove white space after qp_state

Reply:
Done.